graphql/src/store/query.rs

```diff
@@ -2,126 +2,182 @@ use graph::prelude::*;
+        Some(value) => match value {
+            q::Value::Int(n) => Ok(n.as_i64()),
+            _ => Err("first".to_string()),
+        },
+        None => Ok(None),
+    }.map(|n| match n {
+        Some(n) => if n > 0 {
+            Some(n as usize)
+        } else {
+            None
+        },
+        None => None,
+    });
+
+    let skip = match arguments.get(&"skip".to_string()) {
+        Some(value) => match value {
+            q::Value::Int(n) => Ok(n.as_i64()),
+            _ => Err("skip".to_string()),
+        },
+        None => Ok(None),
+    }.map(|n| match n {
+        Some(n) => if n >= 0 {
+            Some(n as usize)
+        } else {
+            None
+        },
+        None => None,
+    });
+
+    if first.is_err() || skip.is_err() {
+        let errors: Vec<String> = vec![first.clone(), skip.clone()]
+            .into_iter()
+            .filter(|r| r.is_err())
+            .map(|e| e.unwrap_err())
+            .collect();
+        return Err(QueryExecutionError::RangeArgumentsError(errors));
+    }
 
-    match (first, skip) {
+    Ok(match (first.unwrap(), skip.unwrap()) {
         (None, None) => None,
         (Some(first), None) => Some(StoreRange { first, skip: 0 }),
         (Some(first), Some(skip)) => Some(StoreRange { first, skip }),
         (None, Some(skip)) => Some(StoreRange { first: 100, skip }),
-    }
+    })
 }
 
 /// Parses GraphQL arguments into a StoreFilter, if present.
 fn build_filter(
     entity: &s::ObjectType,
     arguments: &HashMap<&q::Name, q::Value>,
-) -> Option<StoreFilter> {
-    arguments
-        .get(&"where".to_string())
-        .and_then(|value| match value {
-            q::Value::Object(object) => Some(object),
-            _ => None,
-        }).map(|object| build_filter_from_object(entity, object))
+) -> Result<Option<StoreFilter>, QueryExecutionError> {
+    match arguments.get(&"where".to_string()) {
+        Some(value) => match value {
+            q::Value::Object(object) => Ok(object),
+            _ => return Err(QueryExecutionError::InvalidFilterError),
+        },
+        None => return Ok(None),
+    }.and_then(|object| build_filter_from_object(entity, &object))
 }
 
 /// Parses a GraphQL input object into a StoreFilter, if present.
 fn build_filter_from_object(
     entity: &s::ObjectType,
     object: &BTreeMap<q::Name, q::Value>,
-) -> StoreFilter {
-    StoreFilter::And(
+) -> Result<Option<StoreFilter>, QueryExecutionError> {
+    Ok(Some(StoreFilter::And({
         object
             .iter()
             .map(|(key, value)| {
                 use self::sast::FilterOp::*;
 
                 let (attribute, op) = sast::parse_field_as_filter(key);
 
-                let field = sast::get_field_type(entity, &attribute)
-                    .expect("attribute does not belong to entity");
+                let field = sast::get_field_type(entity, &attribute).ok_or(
+                    QueryExecutionError::EntityAttributeError(
```

Comment:
Can we call this `EntityFieldError` instead of `EntityAttributeError` and also rename the variable `field_name` instead of `attribute`?